### ISSUE TITLE
Inline DTD not processed correctly

### ISSUE THREAD

**Issue by Mingun** [bluk/maybe_xml#9]
You say in documentation:
> The library should scan and evaluate well-formed XML documents correctly. For XML documents which are not well-formed, the behavior is currently undefined. The library does not error when scanning a malformed document.

However when you parse `<!DOCTYPE>` declaration you simply search a `>` outside of quotes (`"` and `'`) and nested `[]`. That is not 100% correct, because `<!DOCTYPE>` declaration can contain XML comments which can have unbalanced `[]`. For example, this is well-formed XML:
```xml
<?xml version="1.1" encoding="UTF-8"?>
<!DOCTYPE root [
<!-- A ] -->
]>
<root/>
```
See the definition of [doctypedecl](https://www.w3.org/TR/xml11/#NT-doctypedecl). It seems that there is not other way to correctly skip DTD other than parse it, but because you guarantee correct work only for well-formed XML documents, you could just skip all `]` inside `<>` (in addition to `"` and `'`). this is possible because all content between `[]` in DTD should be either mark-up (`<!...>`), spaces or _parsed entity reference_ which does not permit `]`.

**Comment by bluk** [comment]
Thanks for the report.

I'm exploring using a parser based on the grammar. I started the changes on the unstable [`parser`](https://github.com/bluk/maybe_xml/tree/parser) branch.

**Comment by bluk** [comment]
This should be fixed in version 0.11.0. The parsing code has been replaced.

By default, the parser is strict. In a future version, there may be options to allow a fast path when the document is known to be XML.

Thanks!
